### PR TITLE
feat(ui): allow run individual tests/suites from the UI

### LINKED ISSUE #6638: [UI] Allow running individual tests/suites
### Clear and concise description of the problem

At the moment, Vitest UI allows re-running all tests in a file.
However, it is commonly useful to re-run just a single test or a suite of tests.
The option to do so is currently disabled in the UI (as this capability is not yet implemented).

### Suggested solution

On however on a test/suite in the sidebar, show a button to re-run a given test/suite.
This would be similar functionality to the "Testing" panel in VS Code.

### Alternative

- Run all tests in a file (slow if file has many tests)
- Stop Vitest, and re-run it with `-t 'somePattern'` argument passed to run only matching tests (clunky)

### Additional context

Original discussion: https://github.com/vitest-dev/vitest/pull/5948#discussion_r1786614993

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by userquin**
### Description

**We need this PR https://github.com/vitest-dev/vitest/pull/6405**

Since we don't have the full tree in the state (missing parent in task), the client will send the task id, the children ids and all parent ids: parent ids are required to avoid mark the test not being run as skipped and so their state will not change.

We also need to reset `configOverride.testNamePattern` when re-running some file (including all files), otherwise we can only re-run previous test/suite run, maybe we can have some side effect here (I need to do some additional tests).

>[!CAUTION]
> Right now the command line will skip other tests/suites not matching the pattern (using t + regexp): the problem with the UI is about enabling/disabling re-run button, disabled when test/suite without state or state with `skip` or `todo` value (check _command line filter behaviour_ screenshot below)

>[!WARNING]
> We should add the `test "path"` when filtering tests/suites (check _filtering from command line_ screenshot below)

resolves #6638

<details>
<summary><i>filtering from command line</i></summary>

imagen
</details>

<details>
<summary><i>command line filter behaviour</i></summary>

imagen
</details>

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/ui/client/composables/client/index.ts [hunk 1]
```diff
@@ -68,6 +69,21 @@ export function runAll() {
   return runFiles(client.state.getFiles()/* , true */)
 }
 
+function clearTaskResult(patterns: string[], task: Task) {
```
Comment (userquin) [inline packages/ui/client/composables/client/index.ts:72]: rename `patterns` to `ids`

Comment (userquin) [reply packages/ui/client/composables/client/index.ts]: not required

File: packages/vitest/src/node/core.ts [hunk 2]
```diff
@@ -688,11 +693,44 @@ export class Vitest {
...
+    }
+  }
+
+  async rerunTestOrSuite(id: string, filename: string) {
+    const patterns: string[] = []
+    const task = this.state.idMap.get(id)
+    let trigger = 'rerun test file'
+    if (task) {
+      trigger = this.isSuite(task) ? 'rerun suite' : 'rerun test'
+      // this.collectTasks(task, patterns)
```
Comment (userquin) [inline packages/vitest/src/node/core.ts:717]: cleanup this

File: packages/ui/client/components/explorer/ExplorerItem.vue [hunk 3]
```diff
@@ -219,12 +238,12 @@ const projectNameTextColor = computed(() => {
...
         v-if="!isReport"
-        v-tooltip.bottom="'Run current test'"
+        v-tooltip.bottom="runButtonTitle"
         data-testid="btn-run-test"
-        title="Run current test"
+        :title="runButtonTitle"
         icon="i-carbon:play-filled-alt"
         text-green5
-        :disabled="type !== 'file'"
+        :disabled="state === 'todo'"
```
Comment (userquin) [inline packages/ui/client/components/explorer/ExplorerItem.vue:246]: review this including `standalone` check, we shouldn't allow run skipped tests (check `ui.spec.ts` and the `standalone` suite: all files when using `standalone` initialized with `state` set to `skip`)

File: packages/ui/client/components/explorer/ExplorerItem.vue [hunk 4]
```diff
@@ -68,12 +69,22 @@ function toggleOpen() {
 }
 
 async function onRun(task: Task) {
+  if (state === 'todo') {
```
Comment (sheremet-va) [inline packages/ui/client/components/explorer/ExplorerItem.vue:72]: state can be dynamic (you can call `ctx.skip`), so let's never assume that the `result.state` dictates if we can attempt to run it

it is just a reported state, it shouldn't block an execution attempt

File: packages/ui/client/components/explorer/ExplorerItem.vue [hunk 5]
```diff
@@ -68,12 +69,22 @@ function toggleOpen() {
...
+  }
+
   onItemClick?.(task)
   if (coverageEnabled.value) {
     disableCoverage.value = true
     await nextTick()
   }
-  await runFiles([task.file])
+
+  if (type === 'file') {
```
Comment (sheremet-va) [inline packages/ui/client/components/explorer/ExplorerItem.vue:82]: Can't we also run folders?

File: packages/ui/client/components/explorer/ExplorerItem.vue (hunk 3 above)
Comment (sheremet-va) [reply packages/ui/client/components/explorer/ExplorerItem.vue:246]: we should ignore any state, `disabled` should never be `true` here

File: packages/vitest/src/api/setup.ts [hunk 6]
```diff
@@ -70,8 +70,11 @@ export function setup(ctx: Vitest, _server?: ViteDevServer) {
           }
           return fs.writeFile(id, content, 'utf-8')
         },
-        async rerun(files) {
-          await ctx.rerunFiles(files)
+        async rerun(files, resetTestNamePattern) {
+          await ctx.rerunFiles(files, undefined, resetTestNamePattern)
+        },
+        async rerunTestOrSuite(id, filename) {
```
Comment (sheremet-va) [inline packages/vitest/src/api/setup.ts:76]: `testOrSuite` can be shortened to `task`

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/ui/client/components/explorer/ExplorerItem.vue (hunk 5 above)
Comment (userquin) [reply packages/ui/client/components/explorer/ExplorerItem.vue:82]: we don't have folders structure in the ui, we only have the tasks from the rpc call (collect and getTestFiles/getFiles): if the new api has a new folder type we'll need to change the ui

File: packages/vitest/src/node/core.ts [hunk 7]
```diff
@@ -688,11 +693,28 @@ export class Vitest {
...
+    return Object.hasOwnProperty.call(task, 'tasks')
+  }
+
+  async rerunTask(id: string) {
+    const task = this.state.idMap.get(id)
+    if (task) {
+      await this.changeNamePattern(
+        task.name,
+        [task.file.filepath],
+        this.isSuite(task) ? 'rerun suite' : 'rerun test',
```
Comment (sheremet-va) [inline packages/vitest/src/node/core.ts:706]: `task.type === 'suite'` is how you check for suite

File: packages/vitest/src/node/core.ts [hunk 8]
```diff
@@ -688,11 +693,28 @@ export class Vitest {
     await this.report('onWatcherStart', this.state.getFiles(files))
   }
 
+  private isSuite(task: RunnerTask): task is RunnerTestSuite {
+    return Object.hasOwnProperty.call(task, 'tasks')
+  }
+
+  async rerunTask(id: string) {
+    const task = this.state.idMap.get(id)
+    if (task) {
```
Comment (sheremet-va) [inline packages/vitest/src/node/core.ts:702]: maybe we should throw an error if task it not found?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/vitest/src/node/core.ts (hunk 8 above)
Comment (userquin) [reply packages/vitest/src/node/core.ts:702]: yeah, I'll ask you about this... or maybe we can rerun all files or passing the filename back again and rerun all files

Comment (sheremet-va) [reply packages/vitest/src/node/core.ts:702]: I think an error is better

**Review by sheremet-va — DISMISSED** [review]

**Comment by sheremet-va** [comment]
Looks good! There are a few conflicts though

**Comment by sheremet-va** [comment]
Sorry to ask, but can you resolve the conflict again?

**Comment by Karthik0521** [comment]
🏆 You decided: These are the winners of Boosting Cycle #3

**Pull request merged**